### PR TITLE
Obstacle memory data structure changed from vector to map.

### PR DESCRIPTION
**Pull request by jasonimercer**
Addresses inefficient compares.

This is CORE-1669 which may help CORE-1665 and CORE-1516.

### CODE REVIEW

**Comment by jasonimercer** [comment]
Not required for current release (June 20/2015). Software testing looks good. Will collect robot testing over the next few weeks.

@skaynama @p6chen @abencz

**Comment by servos** [comment]
LGTM. Side note, is there documentation with details on the changes you have made? (the time decay and inflation stuff) If not can you add some to the wiki for this repo.

File: costmap_2d/plugins/obstacle_layer.cpp [hunk 1]
```diff
@@ -517,12 +513,12 @@ void ObstacleLayer::forgetfulUpdateBounds(double robot_x, double robot_y, double
 
   // clear old observations (unless within keep radius)
   const double earliest_epoch = time_now - obstacle_lifespan_;
-  std::list<TimeWorldPoint>::iterator it;
-  for (it = time_world_points_.begin(); it != time_world_points_.end();)
```
Comment (skaynama) [inline costmap_2d/plugins/obstacle_layer.cpp]: Was the iterator really never incremented before?! :) Good find.

**Comment by jasonimercer** [comment]
The main changes since the LGTM are:
Clearing the pending deletes from the map (obstacle layer)
Move dynamic algorithm select from global space to be a member of inflation (inflation layer)
Only remember data when you are sure of your location (obstacle layer)

File: costmap_2d/plugins/obstacle_layer.cpp [hunk 2]
```diff
@@ -63,10 +63,14 @@ void ObstacleLayer::onInitialize()
...
   obstacle_keep_radius_ = 0.0;       // meters
-  obstacle_queue_size_ = 0;          // observations
-  obstacle_compare_tolerance_ = 0.0; // meters
   use_forgetful_version_ = true;     // flag
 
+  // Initial pose confidence and threshold before we remember new data
+  // Threshold default in costmap_2d/cfg/ObstaclePlugin.cfg
+  pose_confidence_ = 0;
+  pose_confidence_threshold_ = 1;
+  pose_confidence_sub_ = g_nh.subscribe("slam/scan_match_score", 1, &ObstacleLayer::poseConfidenceCallback, this);
```
Comment (jasonimercer) [inline costmap_2d/plugins/obstacle_layer.cpp]: TODO: move to localization_score and make it a configurable parameter.

**Pull request merged**